arch: arm: Remove unnecessary register preservation in Cortex-R port.

stephanosio opened the pull request:
The interrupt exit and swap service routines for Cortex-R unnecessarily
preserve r0 and lr registers when making function calls using bl instruction.

In case of _IntExit in exc_exit.S, the r0 register containing the caller
mode is preserved at the top, and the lr register can safely be assumed to
have been saved into the system mode stack by the interrupt service routine.

In case of __svc in swap_helper.S, since the function saves lr to the system
mode stack at the top and exits through _IntExit, it is not necessary to
preserve lr register when executing bl instructions.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
arch: arm: Remove unnecessary register preservation in Cortex-R port.

The interrupt exit and swap service routines for Cortex-R
unnecessarily preserve r0 and lr registers when making function calls
using bl instruction.

In case of _IntExit in exc_exit.S, the r0 register containing the
caller mode is preserved at the top, and the lr register can safely be
assumed to have been saved into the system mode stack by the interrupt
service routine.

In case of __svc in swap_helper.S, since the function saves lr to the
system mode stack at the top and exits through _IntExit, it is not
necessary to preserve lr register when executing bl instructions.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg commented on arch/arm/core/swap_helper.S:
```diff
@@ -560,9 +560,7 @@ demux:
     beq _oops
 
 #if CONFIG_IRQ_OFFLOAD
-    push {r0, lr}
```
So the LR value at this point is never needed in the SCV handler routine?

ioannisg replied:
I am not really fond of skipping the push/pop sequence around a BL instruction, this seems prone to errors.
If you still want to do that, I think you probably need to write down the full argument as comment, that is, _why_ it is not needed to store the LR. 

/* exception return is done in _IntExit() */ does not reveal the whole story

stephanosio replied:
> I am not really fond of skipping the push/pop sequence around a BL instruction, this seems prone to errors.

As long as lr is preserved (i.e. lr at the time of entering and exiting ISR is the same), there should be no problem. In fact, isr_wrapper is already doing exactly what I intend to do here:
https://github.com/zephyrproject-rtos/zephyr/blob/8c4376e29000cdfbbbe781a40c3977e30dddac0d/arch/arm/core/isr_wrapper.S#L63-L69

stephanosio replied:
> So the LR value at this point is never needed in the SCV handler routine?

For Cortex-R, __svc saves the value of lr into the system mode stack and _IntExit restores this from the system stack into lr, before returning from SVC or IRQ mode.

https://github.com/zephyrproject-rtos/zephyr/blob/8c4376e29000cdfbbbe781a40c3977e30dddac0d/arch/arm/core/swap_helper.S#L519-L534

ioannisg replied:
Exactly. Could you add some inline documentation for that?

stephanosio replied:
> Exactly. Could you add some inline documentation for that?

Add a new commit addressing this: https://github.com/zephyrproject-rtos/zephyr/pull/19389/commits/bc50dffd06699ab17353b8e73c723209c112ea3e

stephanosio commented:
@bbolen Since you are the author of the Cortex-R port, can you help validate this PR?

stephanosio commented:
@ioannisg @bbolen My reasoning is as follows:

* `_IntExit` is called if and only if the processor is executing in either IRQ or SVC mode.
* The only vector to IRQ mode is `_isr_wrapper`, and this function always saves r0-r3+r12+lr_irq to the system mode stack when called.
* The only vector to SVC mode is `__svc`, and this function always saves r0-r3+r12+lr_svc to the system mode stack when called.
* IRQ and SVC modes can only be exited through `_IntExit`.
* `_IntExit` always restores r0-r3+r12+lr_irq/lr_svc before exiting IRQ or SVC mode.

In summary, since lr_irq/lr_svc is always saved when entering, and restored when exiting either IRQ or SVC mode by the functions I listed above, unless the link register-modifying branch instruction is executing within a leaf function, it is not necessary to save lr whenever such a branch instruction is used.

_p.s. Technically speaking, `z_ExcExit` (`_IntExit`) can be called by the FIQ vector `__nmi` which, by default, calls `z_SysNmiOnReset` that locks up the system up and never returns. If a different NMI handler is ever installed, the handler must save the current executing context as `_isr_wrapper` does; otherwise, the system stack will get completely messed up by `z_ExcExit`._

stephanosio committed:
arch: arm: Add Cortex-R exception handling documentation.

Add in-line documentation describing the process of register
preservation and exception handling on Cortex-R.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg commented:
@bbolen could you also take a look?

bbolen commented:
@ioannisg yes, it may be a couple of days before I can get my platform back though to test it.

ioannisg commented:
@bbolen ping ;)

bbolen reviewed (approved):
I haven't been able to test it yet, but the changes themselves look good to me with respect to functionality.

bbolen commented:
For what it's worth, I was able to test out this change and it works on my platform.

stephanosio commented:
I have resolved merge conflicts from recent symbol name changes (e.g. __pendsv to z_arm_pendsv).
